clients/wallet/src/stellar_wallet.rs

```diff
@@ -374,12 +378,113 @@ impl StellarWallet {
 			operations,
 		)?;
 
-		let _ = self.cache.save_tx_envelope(envelope.clone())?;
-
 		self.submit_transaction(envelope).await
 	}
 }
 
+pub enum Tx {
+	V0(TransactionV0),
+	V1(Transaction),
+}
+
+// Error handling methods for StellarWallet.
+impl StellarWallet {
+	pub async fn handle_error(&self, error: Error) -> Result<TransactionResponse, Error> {
+		match &error {
+			Error::HorizonSubmissionError { title: _, status: _, reason, envelope_xdr } => {
+				let transaction_envelope;
+				if let Some(env_xdr) = envelope_xdr {
+					transaction_envelope =
+						TransactionEnvelope::from_base64_xdr(env_xdr).expect("should convert ok");
+					match &reason[..] {
+						"tx_bad_seq" =>
+							self.bump_sequence_number_and_submit(transaction_envelope).await,
+						"tx_internal_error" => self.submit_transaction(transaction_envelope).await,
+						_ => Err(error),
+					}
+				} else {
+					Err(error)
+				}
+			},
+
+			Error::CacheError(CacheError {
+				kind: CacheErrorKind::SequenceNumberAlreadyUsed,
+				path: _,
+				envelope,
+				sequence_number,
+			}) => {
+				if let Some(transaction_envelope) = envelope {
+					return self.bump_sequence_number_and_submit(transaction_envelope.clone()).await
+				}
+				Err(error)
+			},
+			_ => Err(error),
+		}
+	}
+
+	fn get_transaction_from_envelope(&self, tx_envelope: TransactionEnvelope) -> Option<Tx> {
+		match tx_envelope {
+			TransactionEnvelope::EnvelopeTypeTxV0(transaction) => Some(Tx::V0(transaction.tx)),
+			TransactionEnvelope::EnvelopeTypeTx(transaction) => Some(Tx::V1(transaction.tx)),
+			TransactionEnvelope::EnvelopeTypeTxFeeBump(fee_bump_transaction) =>
+				match fee_bump_transaction.tx.inner_tx {
+					FeeBumpTransactionInnerTx::EnvelopeTypeTx(transaction) =>
+						Some(Tx::V1(transaction.tx)),
+					_ => None,
+				},
+			_ => None,
+		}
+	}
+
+	fn update_transaction_sequence_number(&self, tx: Tx) -> Tx {
+		match tx {
+			Tx::V0(transaction) => {
+				let mut updated_tx = transaction;
+				updated_tx.seq_num = updated_tx.seq_num + 1;
+				Tx::V0(updated_tx)
+			},
+			Tx::V1(transaction) => {
+				let mut updated_tx = transaction;
+				updated_tx.seq_num = updated_tx.seq_num + 1;
+				Tx::V1(updated_tx)
+			},
+		}
+	}
+
+	fn get_transaction_envelope(&self, tx: Tx) -> TransactionEnvelope {
+		match tx {
+			Tx::V0(transaction) => self.transaction_v0_into_transaction_envelope(transaction),
+			Tx::V1(transaction) => transaction.into_transaction_envelope(),
+		}
+	}
+
+	pub fn transaction_v0_into_transaction_envelope(
```

Comment:
Maybe instead of duplicating this function here, we could transform the `v0` tx into a `v1` transaction? It seems like it implements the `From` conversion [here](https://github.com/pendulum-chain/substrate-stellar-sdk/blob/d1dd579cc27c614abe9eab7d0cdeae12ea735c0b/src/xdr/impls/transaction.rs#L52) so maybe a simple `into()` would do the trick.

Reply:
Good idea @ebma, I've refactored the code.